### PR TITLE
Add examples benchmarking different contractor settings provided by cotengra

### LINKED ISSUE #212: Add examples benchmarking different contractor settings provided by cotengra
## Task description

Add an example script in `/examples` or a jupyter notebook in `/docs/source/tutorials` to demonstrate the usage of different cotengra optimizers and benchmark the performance metric (pathfinding time, and the metrics for the tensor contraction), specifically with the newly added `SimulatedAnnealing` optimizer in cotengra (accessed by `opt = HyperOptimizer(simulated_annealing_opts={})`, might try combination of different path finding strategy if allowed ).

## Implementation

could try on different circuit sets: random circuits, qft circuits `tc.templates.blocks.qft`, QAOA circuits on different types of graph instances, etc.

## References

https://github.com/jcmgray/cotengra
https://cotengra.readthedocs.io/en/latest/advanced.html#simulated-annealing-simulated-annealing-opts

### PR DESCRIPTION
**Pull request by king-p3nguin**
Resolve #212 

(I am contributing to this project as a unitaryHACK participant)

### CODE REVIEW

File: examples/cotengra_setting_bench.py [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+        loss = 0.0
+        for e in g.edges:
+            loss += c.expectation_ps(z=[e[0], e[1]])
+
+        return K.real(loss)
+
+    params = K.implicit_randn(shape=[nlayers, 2])
+
+    # run only once to trigger the compilation
+    K.jit(
```
Comment (refraction-ray) [inline examples/cotengra_setting_bench.py:62]: why using jit and grad here?

File: examples/cotengra_setting_bench.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+    c = tc.Circuit(n)
+    for i in range(n):
+        c.H(i)
+    for j in range(nlayers):
+        c = tc.templates.blocks.QAOA_block(c, g, param[j, 0], param[j, 1])
+    return c
+
+
+def trigger_cotengra_optimization(n, nlayers, d):
+    g = nx.random_regular_graph(d, n)
```
Comment (refraction-ray) [inline examples/cotengra_setting_bench.py:45]: please consider several different types of graph (say 1D, 2D lattice or all-to-all connection) for the circuit architectures for a better benchmark?

File: examples/cotengra_setting_bench.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+            progbar=True,
+        )
+    else:
+        return ctg.HyperOptimizer(
+            methods=method,
+            optlib=optlib,
+            minimize="flops",
+            parallel=True,
+            max_time=30,
+            max_repeats=30,
```
Comment (refraction-ray) [inline examples/cotengra_setting_bench.py:117]: max_repeats is too small? maybe try 64 or 128. minimize can also be given as "combo" or "size" apart from "flops"

Comment (king-p3nguin) [reply examples/cotengra_setting_bench.py]: I increased the `max_repeats` and added `minimize_args` to benchmark various `minimize` arguments that corengra supports.

File: examples/cotengra_setting_bench.py (hunk 2 above)
Comment (king-p3nguin) [reply examples/cotengra_setting_bench.py]: I added `graph_args` to benchmark the performance for 1D, 2D lattice and all-to-all connected graph.

File: examples/cotengra_setting_bench.py (hunk 1 above)
Comment (king-p3nguin) [reply examples/cotengra_setting_bench.py:61]: Actually, I find that jitting the function here speeds up the path-finding time. As you pointed out, the gradient calculation was unnecessary, so I removed it.

**Comment by refraction-ray** [comment]
thanks for the nice contribution. merged with some small tweaks:
1. increase the max_time to 60
2. comment most of the options by default
3. use `reuse=False` in expectation calculation to directly compute the tensor network for the observable

**Comment by refraction-ray** [comment]
> I find that jitting the function here speeds up the path-finding time

this part still requires further investigation in the future as the path finding stage seems to be irrelevant from jitting things as they are implemented outside the ML backend.
